akka-stream-tests/src/test/scala/akka/stream/scaladsl/UnfoldResourceAsyncSourceSpec.scala

```diff
@@ -0,0 +1,246 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import java.io._
+
+import akka.Done
+import akka.actor.{ NoSerializationVerificationNeeded, ActorSystem }
+import akka.stream.ActorAttributes._
+import akka.stream.Supervision._
+import akka.stream.{ ActorMaterializer, _ }
+import akka.stream.impl.StreamSupervisor.Children
+import akka.stream.impl.{ ActorMaterializerImpl, StreamSupervisor }
+import akka.stream.testkit.TestSubscriber
+import akka.stream.testkit.Utils._
+import akka.stream.testkit.scaladsl.TestSink
+import akka.util.{ ByteString, Timeout }
+import akka.testkit.AkkaSpec
+
+import scala.concurrent.{ Future, Promise }
+import scala.concurrent.duration._
+import scala.util.control.NoStackTrace
+
+class UnfoldResourceAsyncSourceSpec extends AkkaSpec(UnboundedMailboxConfig) {
+
+  val settings = ActorMaterializerSettings(system).withDispatcher("akka.actor.default-dispatcher")
+  implicit val materializer = ActorMaterializer(settings)
+
+  val manyLines = {
+    ("a" * 100 + "\n") * 10 +
+      ("b" * 100 + "\n") * 10 +
+      ("c" * 100 + "\n") * 10 +
+      ("d" * 100 + "\n") * 10 +
+      ("e" * 100 + "\n") * 10 +
+      ("f" * 100 + "\n") * 10
+  }
+  val manyLinesArray = manyLines.split("\n")
+
+  val manyLinesFile = {
+    val f = File.createTempFile("blocking-source-async-spec", ".tmp")
+    new FileWriter(f).append(manyLines).close()
+    f
+  }
+
+  val ex = new RuntimeException() with NoStackTrace
+
+  case object Open extends NoSerializationVerificationNeeded
+  case class Close(r: BufferedReader) extends NoSerializationVerificationNeeded
+  case class Read(r: BufferedReader) extends NoSerializationVerificationNeeded
+
+  val open: () ⇒ Future[BufferedReader] = () ⇒ Promise.successful(new BufferedReader(new FileReader(manyLinesFile))).future
+  val read: (BufferedReader) ⇒ Future[Option[String]] = reader ⇒ Promise.successful(Option(reader.readLine())).future
+  val close: (BufferedReader) ⇒ Future[Done] =
+    reader ⇒ {
+      reader.close()
+      Promise.successful(Done).future
+    }
+
+  "Unfold Resource Async Source" must {
+    "read contents from a file" in assertAllStagesStopped {
+      val createPromise = Promise[BufferedReader]()
+      val readPromise = Promise[Option[String]]()
+      val closePromise = Promise[Done]()
+
+      val resource = new BufferedReader(new FileReader(manyLinesFile))
+      val p = Source.unfoldResourceAsync[String, BufferedReader](() ⇒ {
+        testActor ! Open
+        createPromise.future
+      },
+        reader ⇒ {
+          testActor ! Read(reader)
+          readPromise.future
+        },
+        reader ⇒ {
+          testActor ! Close(reader)
+          closePromise.future
+        })
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[String]()
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      sub.request(1)
+      expectMsg(Open)
+      c.expectNoMsg(200.millis)
```

Comment:
So, this is testing that we do not get an element before the createPromise is completed, that could then be done with a latch/promise in reader to signal that execution went there and checked without any timeout.